1st Class Struct Block Assignments

CarolEidt opened the pull request:
Change block ops to assignments, with block nodes (GT_BLK, GT_OBJ and GT_DYN_BLK) as the lhs, and with GT_STORE_* in the backend.
For this initial change, existing behavior is preserved as much as possible, with a few differences in lea generation for SIMD types.
This causes pessimization in some areas; those, as well as the additional opportunities that can be enabled after this change, have all been marked TODO-1stClassStructs.

CarolEidt commented:
This is the IR change for 1st Class Structs, but largely preserves existing behavior.
I have run diffs on both desktop and x64/ux, and have run the 22 arm tests on my pi (of which VectorUnused_ro fails both with and without these changes).

CarolEidt commented:
@dotnet/jit-contrib PTAL

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -925,9 +925,11 @@ struct GenTree
+    0x40000000 // GT_ASG, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_DYNBLK -- is a volatile block operation
+#define GTF_BLK_UNALIGNED                                                                                              \
+    0x02000000 // GT_ASG, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_DYNBLK -- is an unaligned block operation
+#define GTF_BLK_INIT 0x01000000 // GT_ASG, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_DYNBLK -- is an init block operation
```
This formatting is really weird.

CarolEidt replied:
Hmmm ... I ran jit-format over this, but then did another merge/rebase after that, so perhaps it got messed up.

---
In reply to: [77247280](https://github.com/dotnet/coreclr/pull/7027#discussion_r77247280) [](ancestors = 77247280)

BruceForstall replied:
or maybe jit-format is just that weird.

---
In reply to: [77249318](https://github.com/dotnet/coreclr/pull/7027#discussion_r77249318) [](ancestors = 77249318,77247280)

adiaaida replied:
Yeah, jit-format will do things like this unless you manually split the comment. We can either just go with what jit-format did, or you can manually fix it.

CarolEidt replied:
I manually fixed the comment, and then re-ran jit-format. It looks better now.

---
In reply to: [77255368](https://github.com/dotnet/coreclr/pull/7027#discussion_r77255368) [](ancestors = 77255368)

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -1138,18 +1130,18 @@ struct GenTree
 
-    bool OperIsBlk() const
+    bool OperIsBlk()
     {
```
Is there a reason this can't be 'const' anymore?

CarolEidt replied:
I wouldn't think so. I had to change some methods that pertain to the blk ops/assignments because they had to check some things that aren't const, but this one shouldn't be a problem. I'll change it.

---
In reply to: [77247543](https://github.com/dotnet/coreclr/pull/7027#discussion_r77247543) [](ancestors = 77247543)

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -1821,7 +1817,10 @@ struct GenTree
+    // Returns the i'th child of a dynamic block node, in execution order.
+    GenTreePtr* GetDynBlkOperandInOrder(int index);
+#endif
     // Returns an iterator that will produce the use edge to each operand of this node. Differs
```
Do you need to check in `#if 0` code?

CarolEidt replied:
Oops; thought I'd cleaned those up.

---
In reply to: [77247862](https://github.com/dotnet/coreclr/pull/7027#discussion_r77247862) [](ancestors = 77247862)

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3887,186 @@ struct GenTreeIndir : public GenTreeOp
+
+    // The size of the buffer to be copied.
+    unsigned Size()
+    {
```
const

CarolEidt replied:
Done.

---
In reply to: [77248111](https://github.com/dotnet/coreclr/pull/7027#discussion_r77248111) [](ancestors = 77248111)

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3887,186 @@ struct GenTreeIndir : public GenTreeOp
+
+    // Return true iff the object being copied contains one or more GC pointers.
+    bool HasGCPtr();
+
```
const?

CarolEidt replied:
This can't be `const` unless we make `AsObj()` `const`, and I think that would be a more significant change.

---
In reply to: [77248169](https://github.com/dotnet/coreclr/pull/7027#discussion_r77248169) [](ancestors = 77248169)

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3887,186 @@ struct GenTreeIndir : public GenTreeOp
+    } gtBlkOpKind;
+
+    bool gtBlkOpGcUnsafe;
+
```
Re: naming -- you've moved from "BlkOp" to just "Blk" many places. Should that be true here, also? E.g., BlkKindInvalid?

CarolEidt replied:
Historically "BlkOp" meant the copy or init operation, which I've attempted to preserve, and I've tried to use just "Blk" for the nodes that represent block or struct objects.

---
In reply to: [77248376](https://github.com/dotnet/coreclr/pull/7027#discussion_r77248376) [](ancestors = 77248376)

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3887,186 @@ struct GenTreeIndir : public GenTreeOp
 };
 
+// gtDynBlk  -- 'dynamic block' (GT_DYN_BLK). */
+struct GenTreeDynBlk : public GenTreeBlk
```
IMO, It would be useful for these header comments to be more descriptive about the characteristics of each of the types of block objects.

BruceForstall commented on src/jit/codegenlegacy.cpp:
```diff
@@ -8973,47 +8960,84 @@ void CodeGen::genCodeForCopyObj(GenTreePtr tree, regMaskTP destReg)
-    GenTreePtr op2     = tree->gtGetOp2();
+    GenTreePtr dest    = tree->gtOp.gtOp1;
+    GenTreePtr src     = tree->gtGetOp2();
     regMaskTP  needReg = destReg;
```
Function header?

BruceForstall replied:
I guess this is codegenlegacy... so maybe not worth it

---
In reply to: [77252985](https://github.com/dotnet/coreclr/pull/7027#discussion_r77252985) [](ancestors = 77252985)

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -2747,61 +2745,69 @@ void CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
-        case GT_COPYBLK:
+        case GT_STORE_DYN_BLK:
+        case GT_STORE_BLK:
         {
```
This function is already huge. Couldn't we refactor this case out as genCodeForStoreBlk()?

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -5279,86 +5323,156 @@ void CodeGen::genConsumePutStructArgStk(
+void CodeGen::genConsumeBlockSize(GenTreeBlk* blkNode, regNumber sizeReg)
+{
+    unsigned blockSize = blkNode->Size();
+    if (sizeReg != REG_NA)
```
put this within the 'if'

CarolEidt replied:
Actually, I think there may be a bug here - in the `GT_STORE_DYN_BLK` case, it should probably be `genConsumeRegAndCopy(blkNode->AsDynBlk()->gtDynamicSize, sizeReg)`. This probably hasn't been hit because 1) the dynamic size is not common and 2) it is probably usually already in the right register.

---
In reply to: [77254555](https://github.com/dotnet/coreclr/pull/7027#discussion_r77254555) [](ancestors = 77254555)

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -5279,86 +5323,156 @@ void CodeGen::genConsumePutStructArgStk(
+}
+
+//------------------------------------------------------------------------
+// genConsumeBlockDst: Ensure that the block source address is in its
```
>genConsumeBlockDst [](start = 3, length = 18)

copy-and-paste error: genConsumeBlockDst

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -5279,86 +5323,156 @@ void CodeGen::genConsumePutStructArgStk(
+// Arguments:
+//    blkNode - The block node
+//
+
```
comment should include "Return Value:"

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -5279,86 +5323,156 @@ void CodeGen::genConsumePutStructArgStk(
+            // This must be a local.
+            assert(src->OperIsLocal());
+            return nullptr;
+        }
```
We don't genConsumeReg() in this case?

CarolEidt replied:
No, in this case it is doing a block copy from a local, which (in this case) must be on the stack, and therefore has no associated register. I'll add a comment to that effect.

---
In reply to: [77255825](https://github.com/dotnet/coreclr/pull/7027#discussion_r77255825) [](ancestors = 77255825)

BruceForstall commented on src/jit/compiler.cpp:
```diff
@@ -4428,11 +4428,12 @@ void Compiler::compCompile(void** methodCodePtr, ULONG* methodCodeSize, CORJIT_F
-            for (GenTreeStmt* statement = block->firstStmt(); statement != nullptr; statement = statement->getNextStmt())
+            for (GenTreeStmt* statement = block->firstStmt(); statement != nullptr;
+                 statement              = statement->getNextStmt())
             {
```
OMG, this formatting is awful :(

CarolEidt replied:
I'm not a fan of this either - is there a way that I can improve it that jit-format will be happy with?

---
In reply to: [77255945](https://github.com/dotnet/coreclr/pull/7027#discussion_r77255945) [](ancestors = 77255945)

adiaaida replied:
We can change what jit-format does for assignments and declarations. I've been hearing a lot of people say they don't like how either instance is formatted.

adiaaida replied:
Otherwise, rename statement to stmt (or something similar?) to make the for statement fit in < 120 characters.

CarolEidt replied:
OK, I'll rename to stmt for now - and leave to others the broader question ;-)

---
In reply to: [77263018](https://github.com/dotnet/coreclr/pull/7027#discussion_r77263018) [](ancestors = 77263018)

BruceForstall commented on src/jit/compiler.hpp:
```diff
@@ -4702,11 +4702,9 @@ inline bool BasicBlock::endsWithTailCallConvertibleToLoop(Compiler* comp, GenTre
-    gtBlockOpInit(result, GT_COPYOBJ, source->Dest(), source->Source(), source->ClsTok(), source->IsVolatile());
-    return result;
+    return gtClone(source);
 }
```
Why not just replace all uses of gtCloneCpObjNode() and then get rid of this function?

CarolEidt replied:
OK

---
In reply to: [77256154](https://github.com/dotnet/coreclr/pull/7027#discussion_r77256154) [](ancestors = 77256154)

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -17719,6 +17724,52 @@ void Compiler::fgSetTreeSeqHelper(GenTreePtr tree, bool isLIR)
+    // Special handling for dynamic block ops.
+    if (((tree->OperGet() == GT_ASG) && (tree->gtGetOp1()->OperGet() == GT_DYN_BLK)) ||
+        (tree->OperGet() == GT_STORE_DYN_BLK))
+    {
```
Should you have a tree->IsDynBlk() function (or similar) to encapsulate this?

CarolEidt replied:
Actually, there *is* an `OperIsDynBlkOp`, but apparently I didn't use it.

---
In reply to: [77256474](https://github.com/dotnet/coreclr/pull/7027#discussion_r77256474) [](ancestors = 77256474)

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22113,14 +22157,11 @@ GenTreePtr      Compiler::fgInlinePrependStatements(InlineInfo* inlineInfo)
-                                          gtNewIconNode(info.compCompHnd->getClassSize(structType)), // Size
-                                          false);           // volatil
+                                          info.compCompHnd->getClassSize(structType), // Size
+                                          false,            // volatil
```
>volatil [](start = 63, length = 7)

typo: volatil

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -911,6 +927,27 @@ Compiler::fgWalkResult Compiler::fgWalkTreePostRec(GenTreePtr* pTree, fgWalkData
+            }
+            break;
+#if 0
         case GT_PHI:
```
GT_PHI is under `#if 0` now?

CarolEidt replied:
Hmm ... I don't recall making that change, or why I would have done so. It's just before the place where the old block ops were so maybe I had previously ifdef'd them out and the `#if 0` migrated. I'll remove the #if 0 (wonder why it passes all the tests, though?)

---
In reply to: [77257978](https://github.com/dotnet/coreclr/pull/7027#discussion_r77257978) [](ancestors = 77257978)

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -2516,6 +2573,21 @@ unsigned Compiler::gtHashValue(GenTree* tree)
+
+                case GT_OBJ:
+                case GT_STORE_OBJ:
+                    hash ^= reinterpret_cast<unsigned>(tree->AsObj()->gtClass);
```
is there a reason this is `^` instead of `+` like every other case?

CarolEidt replied:
This is the same as the previous `GT_OBJ` handling - though your comment made me realize that I'd failed to remove `GT_OBJ` from the unary case above. I'll remove that.

---
In reply to: [77258271](https://github.com/dotnet/coreclr/pull/7027#discussion_r77258271) [](ancestors = 77258271)

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -6586,115 +6803,167 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
+    {
+        newBlkOrObjNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd, size);
+    }
     // An Obj is not a global reference, if it is known to be a local struct.
```
why is there a scope here?

CarolEidt replied:
Because I failed to eliminate it when I got rid of a condition. Fixed.

---
In reply to: [77258711](https://github.com/dotnet/coreclr/pull/7027#discussion_r77258711) [](ancestors = 77258711)

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -6814,105 +7077,113 @@ void Compiler::gtBlockOpInit(
+        srcOrFillVal->gtFlags |= GTF_DONT_CSE;
+        if (srcOrFillVal->OperIsIndir() && (srcOrFillVal->gtGetOp1()->gtOper == GT_ADDR))
+        // && (srcOrFillVal->gtGetOp1()->gtGetOp1()->TypeGet() == dst->TypeGet()))
+        {
```
why commented out?

CarolEidt replied:
Because it turns out that the existing code just retypes the source in that case - but I failed to eliminate this after successfully matching the former behavior.

---
In reply to: [77258955](https://github.com/dotnet/coreclr/pull/7027#discussion_r77258955) [](ancestors = 77258955)

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -7619,6 +7881,14 @@ GenTreePtr Compiler::gtCloneExpr(GenTree* tree,
+                                                  gtCloneExpr(tree->gtDynBlk.gtDynamicSize, addFlags, varNum, varVal));
+            // ??? WHY IS THIS HERE?? THE ABOVE SHOULD HAVE SET IT CORRECTLY ???
+            // copy->gtDynBlk.gtOp1 = tree->gtDynBlk.gtOp1;
+            break;
```
remove? change to TODO?

CarolEidt replied:
Another missed cleanup. I wasn't sure why I'd added it, so I commented it out and then failed to remove it.

---
In reply to: [77259059](https://github.com/dotnet/coreclr/pull/7027#discussion_r77259059) [](ancestors = 77259059)

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -8235,6 +8527,45 @@ unsigned GenTree::NumChildren()
+    }
+}
+#endif
+
```
delete? (`#if 0`)

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -8244,6 +8575,14 @@ GenTreePtr GenTree::GetChild(unsigned childNum)
+    {
+        return gtOp.gtOp1->AsDynBlk()->gtDynamicSize;
+    }
     else if (OperIsBinary())
```
It's unfortunate that some of these dynblk special cases introduce checks that are probably more expensive than the cheaper and more common cases that follow. Same is true in the GTK_SMPOP check I saw elsewhere that is now preceded by a relatively more expensive condition. I think GetChild is relatively hot, too.

CarolEidt replied:
Yes - I don't think that we actually need to preserve this behavior long-term, but I was really trying to minimize diffs for this initial checkin. But there should be a TODO-1stClassStructs here to revisit this. I'll add one.

---
In reply to: [77259527](https://github.com/dotnet/coreclr/pull/7027#discussion_r77259527) [](ancestors = 77259527)

BruceForstall commented:
There are (obviously) tons of code changes, and I can't say I understood most of them. So my comments were mostly minor, surface-y. But from what I've seen, ... LGTM

CarolEidt commented:
Wow - thanks @BruceForstall and @adiaaida for reviewing this so quickly! I've pushed another commit addressing the comments so far.

---
In reply to: [244223991](https://github.com/dotnet/coreclr/pull/7027#issuecomment-244223991) [](ancestors = 244223991)

sivarv commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3884,198 @@ struct GenTreeIndir : public GenTreeOp
+        else
+        {
+            // ChangeOper(GT_BLK);
+        }
```
Code inside this 'else' block is commented out.  Is that intended?  If yes, it could be deleted?

CarolEidt replied:
Ah, yes - I decided to eliminate that, since it's not really necessary. I will remove it.

---
In reply to: [77271784](https://github.com/dotnet/coreclr/pull/7027#discussion_r77271784) [](ancestors = 77271784)

sivarv commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3884,198 @@ struct GenTreeIndir : public GenTreeOp
+{
+public:
+    // The data to be stored (null for GT_BLK)
+    GenTree*& Data()
```
Is there a reason why return type has to be a reference to GenTree* instead of just GenTree*?  A comment would be useful here why it is needed.

CarolEidt replied:
I am merely following the convention used elsewhere (see `Addr` in the base type - `GenTreeIndir`)

---
In reply to: [77272285](https://github.com/dotnet/coreclr/pull/7027#discussion_r77272285) [](ancestors = 77272285)

sivarv commented on src/jit/gentree.h:
```diff
@@ -5010,12 +5061,11 @@ inline var_types& GenTree::CastToType()
 // Notes:
 //    Of the block ops only GT_COPYOBJ is allowed to have GC pointers.
```
There is no GT_COPYOBJ post this change.  Should it be reading it as GT_OBJ?

CarolEidt replied:
Yes, I'll fix it to say "Of the block nodes, only GT_OBJ and ST_STORE_OBJ are allowed to have GC pointers."

---
In reply to: [77273635](https://github.com/dotnet/coreclr/pull/7027#discussion_r77273635) [](ancestors = 77273635)

sivarv commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3884,198 @@ struct GenTreeIndir : public GenTreeOp
+        BlkOpKindUnroll,
+    } gtBlkOpKind;
+
+    bool gtBlkOpGcUnsafe;
```
As per my understanding a GT_BLK node doesn't contain gc-ptrs and hence whether it is GcUnsafe will always be false.  This makes me wonder, whether this member belongs to GenTreeObj instead of GenTreeBlk?  Essentially, are GenTreeBlk and GenTreeDynBlk nodes unnecessarily paying the space penalty of having gtBlkOpGcUnsafe member?

We add a member method in base class GenTreeBlk::IsGCUnsafe() - that will return false for GT_BLK/DYN_BLK and gtBlkOpGCUnsafe for GT_OBJ.

CarolEidt replied:
This is something that obviously could change in future, but currently `GT_BLK` is a "small" node, and this doesn't push it over the limit. It makes it a bit simpler to check in the block handling code. So, for now I would suggest that we leave this as-is.

---
In reply to: [77274153](https://github.com/dotnet/coreclr/pull/7027#discussion_r77274153) [](ancestors = 77274153)

adiaaida commented on src/jit/assertionprop.cpp:
```diff
@@ -3414,6 +3421,26 @@ GenTreePtr Compiler::optAssertionProp_Ind(ASSERT_VALARG_TP assertions, const Gen
+    // TODO-1stClassStructs: All indirections should be handled here, but
+    // previously, when these indirections were GT_OBJ, or implicit due to a block
+    // copy or init, they were not being handled.
+    bool propagateIndir = true;
```
Is this bool used anywhere?

CarolEidt replied:
Um, no ... I had more complex logic here before, and once I realized that I could simplify it I neglected to remove this. Will delete it.

---
In reply to: [77274805](https://github.com/dotnet/coreclr/pull/7027#discussion_r77274805) [](ancestors = 77274805)

adiaaida commented on src/jit/codegenarm64.cpp:
```diff
@@ -3518,53 +3518,57 @@ void CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
+                genCodeForCpObj(treeNode->AsObj());
+            }
             break;
+            __fallthrough;
```
If we break in the line above, will we ever hit this __fallthrough?

CarolEidt replied:
No - and it's a bit surprising that this wasn't discovered, though I think most of the no-gc-ptr copies get a GT_BLK on the lhs from the beginning.

---
In reply to: [77274977](https://github.com/dotnet/coreclr/pull/7027#discussion_r77274977) [](ancestors = 77274977)

adiaaida commented on src/jit/compiler.h:
```diff
@@ -1877,18 +1877,21 @@ class Compiler
+    GenTree* gtNewBlockVal(GenTreePtr addr, unsigned size);
 
-    GenTreeBlkOp* gtCloneCpObjNode(GenTreeCpObj* source);
+    GenTree* gtNewCpObjNode(GenTreePtr dst, GenTreePtr src, CORINFO_CLASS_HANDLE structHnd, bool volatil);
```
nit: volatil

adiaaida commented on src/jit/flowgraph.cpp:
```diff
@@ -9000,7 +9000,7 @@ void Compiler::fgSimpleLowering()
                         tree->gtPrev = add;
 #else
-                        range.InsertAfter(arr, con, add);
+                            range.InsertAfter(arr, con, add);
```
For the life of me, I cannot figure out why clang-format keeps wanting to indent this. It drives me nuts, and I've seen it on multiple of my own format runs, and I know there's nothing we can do about it right now, but it's just strange.

CarolEidt replied:
I share your pain ;-)

---
In reply to: [77275413](https://github.com/dotnet/coreclr/pull/7027#discussion_r77275413) [](ancestors = 77275413)

sivarv commented on src/jit/gentree.h:
```diff
@@ -4037,26 +3884,198 @@ struct GenTreeIndir : public GenTreeOp
+    void SetData(GenTree* dataNode)
+    {
+        gtOp2 = dataNode;
+    }
```
As per my understanding a gt_blk/dyn_blk/obj nodes represent address and shape formerly on block ops.  This would mean that data will always be null?  Please add a comment here what would gtOp2 contains

CarolEidt replied:
No, it is not always null. These get transformed to `GT_STORE_BLK`, etc. by `Rationalizer`. Note that above it clearly indicates that this node is used for `GT_STORE_BLK` as well (and, by inheritance, by `GT_STORE_OBJ` and `GT_STORE_DYN_BLK`).

---
In reply to: [77278869](https://github.com/dotnet/coreclr/pull/7027#discussion_r77278869) [](ancestors = 77278869)

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6814,105 +7071,112 @@ void Compiler::gtBlockOpInit(
-            LclVarDsc* lclVarDsc             = &lvaTable[lclNum];
-            lclVarDsc->lvUsedInSIMDIntrinsic = true;
+            src = src->AsIndir()->Addr()->gtGetOp1();
+        }
```
I see this logic is already present in caller of this method in gtNewBlkOpNode(). Is this logic really needed here?

CarolEidt replied:
Actually, this is also called by `impAssignStructPtr`, so I think the right answer is probably to eliminate it from `gtNewBlkOpNode`.

---
In reply to: [77280046](https://github.com/dotnet/coreclr/pull/7027#discussion_r77280046) [](ancestors = 77280046)

CarolEidt replied:
On looking at this further, the logic here is simply getting the source for setting the `lvUsedInSIMDIntrinsic` flag, whereas the logic in `gtNewBlkOpNode` is modifying what will be used as the source in the assignment. While there is probably an opportunity for simplification, I'm not inclined to add the SIMD check to `impAssignStructPtr`, which is already rather complex, so I think it's best to leave this here.

---
In reply to: [77283021](https://github.com/dotnet/coreclr/pull/7027#discussion_r77283021) [](ancestors = 77283021,77280046)

briansull commented on src/jit/valuenum.cpp:
```diff
@@ -5162,7 +5213,8 @@ void Compiler::fgValueNumberTree(GenTreePtr tree, bool evalAsgLhsInd)
                                 // the indirection is reading beyond the end of the field
                                 //
-                                lcl->gtVNPair.SetBoth(vnStore->VNForExpr(compCurBB, typ)); // return a new unique value number
+                                lcl->gtVNPair.SetBoth(vnStore->VNForExpr(compCurBB, typ)); // return a new unique value
```
Why did this get reformatted now?
You didn't change anything

CarolEidt replied:
All the formatting got messed up on one of my merges, so I re-ran jit-format. So there are some formatting-only changes mixed in here.

---
In reply to: [77282193](https://github.com/dotnet/coreclr/pull/7027#discussion_r77282193) [](ancestors = 77282193)

briansull commented on src/jit/valuenum.cpp:
```diff
@@ -5858,8 +5916,9 @@ void Compiler::fgValueNumberTree(GenTreePtr tree, bool evalAsgLhsInd)
-                                                                                   fldSeq->m_next, storeVal, indType, compCurBB);
+                                        storeVal =
+                                            vnStore->VNApplySelectorsAssign(VNK_Liberal, fldMapVN, fldSeq->m_next,
+                                                                            storeVal, indType, compCurBB);
```
There quite a few of these reformatting changes, that I would not have expected to see

CarolEidt replied:
Yes, I think there were some commits that didn't have jit-format run on them; sorry about that.

---
In reply to: [77282298](https://github.com/dotnet/coreclr/pull/7027#discussion_r77282298) [](ancestors = 77282298)

briansull commented on src/jit/rationalize.cpp:
```diff
@@ -157,19 +157,20 @@ void Compiler::fgFixupArgTabEntryPtr(GenTreePtr parentCall, GenTreePtr oldArg, G
-// Rewrite InitBlk involving SIMD vector into stlcl.var of a SIMD type.
+// Rewrite a non-TYP_STRUCT indirection as GT_IND(GT_LEA(obj.op1)), or as a simple
+// lclVar if possible.
 //
```
Does this only handle SIMD operations?
If so the comment should say that instead of "non-TYP_STRUCT"

CarolEidt replied:
So ... that comment is somewhat anticipatory - I had made additional changes to use the enregisterable struct types, which also need this transformation. I'll add an additional comment.

---
In reply to: [77282496](https://github.com/dotnet/coreclr/pull/7027#discussion_r77282496) [](ancestors = 77282496)

CarolEidt replied:
Actually, I see that this is something that should be going away eventually, so I'll just make it clear that this is only for SIMD.

---
In reply to: [77282570](https://github.com/dotnet/coreclr/pull/7027#discussion_r77282570) [](ancestors = 77282570,77282496)

briansull commented on src/jit/gentree.h:
```diff
@@ -1459,9 +1454,10 @@ struct GenTree
-            case GT_COPYBLK:
-            case GT_COPYOBJ:
+            case GT_BLK:
+            case GT_OBJ:
```
A comment explaining what Op2 represents when we have GT_BLK or GT_OBJ as I would normally think of these two a GT_IND and thus they don't have an Op2.

CarolEidt replied:
Hmm. I don't think the the non-store variants belong  belong here. I'll remove them.

---
In reply to: [77282926](https://github.com/dotnet/coreclr/pull/7027#discussion_r77282926) [](ancestors = 77282926)

briansull replied:
I think all of the unary ops have a null for op2.


---
In reply to: [77283180](https://github.com/dotnet/coreclr/pull/7027#discussion_r77283180) [](ancestors = 77283180,77282926)

CarolEidt replied:
Yes, but they are not GTK_BINOP, so they won't reach here. However, GT_STOREIND shouldn't be here because it *must* have an op2, which is the data to be stored. So I have deleted it as well as the others that I added.

---
In reply to: [77283459](https://github.com/dotnet/coreclr/pull/7027#discussion_r77283459) [](ancestors = 77283459,77283180,77282926)

briansull commented on src/jit/optimizer.cpp:
```diff
@@ -6936,6 +6936,17 @@ void Compiler::optComputeLoopSideEffectsOfBlock(BasicBlock* blk)
+                    if (!tree->DefinesLocal(this, &lclVarTree, &isEntire))
+                    {
+                        // For now, assume arbitrary side effects on the heap...
+                        // JBTODO [ddetlefs, 11/2012] Why not be complete, and get this case right?
```
Should remove the ddetlefs here...

CarolEidt commented:
@dotnet-bot test this please

sivarv commented on src/jit/gentree.h:
```diff
@@ -4626,24 +4636,58 @@ struct GenTreeAllocObj final : public GenTreeUnOp
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) != 0));
```
I think the condition to be used on 4685 is that the oper is one of STORE_OBJ/BLK/DYN_BLK, then it would eliminate the need to check that Data() != nullptr.

How about adding OperIsStoreBlkOp() and using the same here?

sivarv replied:
Do we need to mark initblk nodes with GTF_BLK_INIT flag?  Can't that be inferred from Data?  If Oper is a StoreBlkOP and Data is a const int, then it must be init blk?  Is there a reason we cannot do it?

CarolEidt replied:
I believe it is more expedient to check that `Data != nullptr` than to check for all the store opcodes.

---
In reply to: [77381269](https://github.com/dotnet/coreclr/pull/7027#discussion_r77381269) [](ancestors = 77381269)

CarolEidt replied:
No because we can, in future, have block nodes for which the types have been modified to a fixed-size type, and the rhs may be a constant. The semantics of initBlk are somewhat unique in that it's not really an assignment of the full rhs, but rather an assignment of the lower 8 bits of the rhs constant expanded to fit the size of the block. I previously did not have the GTK_BLK_INIT flag, but it became messy to check for the cases, so this seems the better solution.

---
In reply to: [77381784](https://github.com/dotnet/coreclr/pull/7027#discussion_r77381784) [](ancestors = 77381784)

sivarv replied:
No it is not.  What I am suggesting is to modify else if check as follows:
Note OperIsStoreBlkOp() return true when oper is GT_STORE_OBJ/BLK/DYN_BLK.

else if (OperIsStoreBlkOp())
{
   return (gtFlags & GTF_BLK_INIT) != 0;
}

Here OperIsBlk() is unnecessarily checking against GT_OBJ/BLK/DYN_BLK and later excluding them by Data() != nullptr, which is not required if we used the above.

---
In reply to: [77387651](https://github.com/dotnet/coreclr/pull/7027#discussion_r77387651) [](ancestors = 77387651,77381269)

CarolEidt replied:
Ah, I see. I missed that you were suggesting to eliminate the `OperIsBlk()` check (rather than just replace the check on Data. I'll make that change.

---
In reply to: [77390490](https://github.com/dotnet/coreclr/pull/7027#discussion_r77390490) [](ancestors = 77390490,77387651,77381269)

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
 
 // Creates a new Obj node.
-GenTreeObj* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr)
+GenTree* Compiler::gtNewObjNode(CORINFO_CLASS_HANDLE structHnd, GenTree* addr)
```
Since you are editing this routine, you might want to add a function header.

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
     var_types nodeType = impNormStructType(structHnd);
     assert(varTypeIsStruct(nodeType));
-    GenTreeObj* objNode = new (this, GT_OBJ) GenTreeObj(nodeType, addr, structHnd);
+    unsigned size = info.compCompHnd->getClassSize(structHnd);
```
Note that impNormStructType() will call getClassSize() and getClassGCLayout() JIT-EE interface methods.
Afterwards, this routine is going to call getClassSize() and later there will be a call to getClassGCLayout() in  gtSetObjGcInfo().

Are we making the same JIT-EE call multiple times on a given structHnd while trying to create a blkop node?

CarolEidt replied:
I don't believe that we are. I originally had the call to `getClassGCLayout()` here, but found that it caused additional JIT-EE calls that did not exist before. I am nearly certain that I have not added any *new* JIT-EE calls (via SuperPmi), but there were already places that were calling the same JIT-EE call multiple times - I don't *think* I have increased that number, but it is more difficult to verify than that we are not making any *new* calls.

---
In reply to: [77384418](https://github.com/dotnet/coreclr/pull/7027#discussion_r77384418) [](ancestors = 77384418)

CarolEidt commented:
@dotnet-bot test this please

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
+void Compiler::gtSetObjGcInfo(GenTreeObj* objNode)
+{
+    CORINFO_CLASS_HANDLE structHnd  = objNode->gtClass;
+    var_types            nodeType   = impNormStructType(structHnd);
```
Note that impNormStructType() is going to all getClassSize() and getClassGCLayout() EE methods.  These could already be called while creating GenTreeObj node and gets called again.  Since nodeType set on GenTreeObj is that of normalized type, cant' we use the type of objNode here?

CarolEidt replied:
Yes, I think we should be able to do that.

---
In reply to: [77385267](https://github.com/dotnet/coreclr/pull/7027#discussion_r77385267) [](ancestors = 77385267)

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
+            var_types simdBaseType; // Dummy argument
+            slots    = (unsigned)(roundUp(size, TARGET_POINTER_SIZE) / TARGET_POINTER_SIZE);
+            gcPtrs   = new (this, CMK_ASTNode) BYTE[slots];
+            nodeType = impNormStructType(structHnd, gcPtrs, &gcPtrCount, &simdBaseType);
```
Here is another impNormStructType() that will lead to getClassSize() and getClassGCLayout() JIT-EE calls.  Are these method calls cheap enough?  Even if cheap, it would be a good idea to avoid calling impNormStructType() multiple times while creating a blkOp node.

CarolEidt replied:
This is the first/only place where we get the actual GC layout for object nodes (the other place we do this for struct types is when we set up the lclVars). Prior to this, it is necessary that we know the type and size, so we have to call `impNormStructType()` without getting the GC layout. We may determine that it is better to pay the full price up-front, but that leads to a different set of JIT-EE calls than we were previously doing.

---
In reply to: [77385804](https://github.com/dotnet/coreclr/pull/7027#discussion_r77385804) [](ancestors = 77385804)

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6586,115 +6799,165 @@ GenTreePtr Compiler::gtNewAssignNode(GenTreePtr dst, GenTreePtr src)
+//
+//  dst        - The target to copy the struct to
+//  src        - The source to copy the struct from
+//  structHnd  - A class token that represents the type of object being copied. May be null
```
It would be a good idea to be more explicit and mention that dst and src are actually addresses.  To be more clear it would be a good idea to rename the arguments as dstAddr and SrcAddr.

CarolEidt replied:
Yes, good point. I had attempted to make this take actual values, but it was difficult to preserve existing behavior. In future this may be changed to take the struct values, but for now I'll do as you suggest and change the names.

---
In reply to: [77387710](https://github.com/dotnet/coreclr/pull/7027#discussion_r77387710) [](ancestors = 77387710)

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6814,105 +7071,112 @@ void Compiler::gtBlockOpInit(
 //
 // Arguments:
-//    oper          - GT_COPYBLK, GT_INITBLK or GT_COPYOBJ
 //    dst           - Destination or target to copy to / initialize the buffer.
```
If I understood correctly, here dst has to be either a GT_OBJ/BLK/DYN_BLK or a LclVar?

If yes, can we be more clear in function header comments and also assert the same in this routine?

CarolEidt replied:
Yes, I believe that's the case. I will add comments and an assert.

---
In reply to: [77388649](https://github.com/dotnet/coreclr/pull/7027#discussion_r77388649) [](ancestors = 77388649)

sivarv commented on src/jit/gentree.cpp:
```diff
@@ -6746,42 +7009,36 @@ void GenTreeIntCon::FixupInitBlkValue(var_types asgType)
-//     - result is a GenTreeBlkOp node that is the node to be initialized.
-//     - oper must be either GT_INITBLK or GT_COPYBLK
+//     - result is an assignment node that is to be initialized.
 //     - dst is the target (destination) we want to either initialize or copy to
```
Here too dst has to be either a GT_OBJ/BLK/DYN_BLK or a LclVar?  If yes, please update both the function header comment and assert the same in this routine.

CarolEidt replied:
No, the lhs may have been modified to a different type. However, I can add a comment and assert that if it is TYP_STRUCT, the lhs must be a block node or lclVar.

---
In reply to: [77389617](https://github.com/dotnet/coreclr/pull/7027#discussion_r77389617) [](ancestors = 77389617)

CarolEidt replied:
Indeed, if it's not TYP_STRUCT, it should do an early return. I'll make that change as well.

---
In reply to: [77396814](https://github.com/dotnet/coreclr/pull/7027#discussion_r77396814) [](ancestors = 77396814,77389617)

sivarv commented on src/jit/gentree.h:
```diff
@@ -4626,24 +4636,58 @@ struct GenTreeAllocObj final : public GenTreeUnOp
-    return (gtOper == GT_INITBLK || gtOper == GT_COPYBLK || gtOper == GT_COPYOBJ);
+    return (((gtOper == GT_ASG) && varTypeIsStruct(gtOp.gtOp1))
+#ifndef LEGACY_BACKEND
+            || (OperIsBlk() && (AsBlk()->Data() != nullptr))
```
It would be more clear if we added OperIsStoreBlkOp() that returns true if oper is GT_STORE_OBJ/BLK/DYN_BLK.

CarolEidt replied:
Will do.

---
In reply to: [77389887](https://github.com/dotnet/coreclr/pull/7027#discussion_r77389887) [](ancestors = 77389887)

sivarv commented on src/jit/importer.cpp:
```diff
@@ -1086,8 +1085,17 @@ GenTreePtr Compiler::impAssignStruct(GenTreePtr           dest,
+    }
+
+    // TODO-1stClassStructs: Avoid creating an address if it is not needed,
+    // or re-creating a Blk node if it is.
```
Not sure I understood this comment. In what cases we can avoid creating an addr/blk node that we are not avoiding?  This comments needs to be expanded with those details or if it not relevant should be deleted.

CarolEidt replied:
In order to preserve existing behavior, it is sometimes necessary to get an address, and "sink" the struct object under an indirection, however it is often not needed downstream. Eliminating these early changes the IR, and thus the generated code (usually for the better, but surprisingly not always). I will expand this comment to make that a bit mroe clear.

---
In reply to: [77394689](https://github.com/dotnet/coreclr/pull/7027#discussion_r77394689) [](ancestors = 77394689)

sivarv commented on src/jit/importer.cpp:
```diff
@@ -14078,7 +14208,20 @@ void Compiler::impImportBlockCode(BasicBlock* block)
+                    op1  = new (this, GT_DYN_BLK) GenTreeDynBlk(op1, op3);
+                    size = 0;
+                }
+                op2 = gtNewOperNode(GT_IND, TYP_STRUCT, op2);
```
Here itself we can check to see whether src is GT_ADDR and if so use its op1 as the value so that we don't unnecessarily create and discard GT_IND node?

CarolEidt replied:
Yes, I think it's safe to eliminate this one.

---
In reply to: [77397229](https://github.com/dotnet/coreclr/pull/7027#discussion_r77397229) [](ancestors = 77397229)

briansull commented on src/jit/morph.cpp:
```diff
@@ -8235,43 +8253,21 @@ void Compiler::fgAssignSetVarDef(GenTreePtr tree)
 
 GenTreePtr Compiler::fgMorphOneAsgBlockOp(GenTreePtr tree)
 {
-    genTreeOps oper = tree->gtOper;
```
This method really could use a method header comment.

sivarv commented:
> //    If we leave it as a GT_COPYBLK or GT_COPYOBJ we will call lvaSetVarDoNotEnregister() on both Source() and Dest()

Nit: To avoid confusion while reading it, you might want to end this sentence with "."

---
Refers to: src/jit/morph.cpp:9198 in 86b72d1. [](commit_id = 86b72d15a213ddc964efc84043e50ec90339ba9b, deletion_comment = False)

sivarv commented on src/jit/morph.cpp:
```diff
@@ -8795,6 +9209,37 @@ GenTreePtr Compiler::fgMorphCopyBlock(GenTreePtr tree)
+
+#if FEATURE_MULTIREG_RET
+    // If this is a multi-reg return, we will not do any morphing of this node.
+    if ((rhs->OperGet() == GT_CALL) && rhs->AsCall()->HasMultiRegRetVal())
```
We can use rhs->IsMultiRegCall() here as if condition.

sivarv commented:
>  *                       [dest]     [val/src]

This comment needs to be edited?

---
Refers to: src/jit/flowgraph.cpp:18337 in 86b72d1. [](commit_id = 86b72d15a213ddc964efc84043e50ec90339ba9b, deletion_comment = False)

sivarv commented on src/jit/rationalize.cpp:
```diff
@@ -157,19 +157,20 @@ void Compiler::fgFixupArgTabEntryPtr(GenTreePtr parentCall, GenTreePtr oldArg, G
 // Arguments:
-//    ppTree      - A pointer-to-a-pointer for the GT_INITBLK
+//    ppTree      - A pointer-to-a-pointer for the GT_BLK or GT_OBJ
 //    fgWalkData  - A pointer to tree walk data providing the context
```
Arguments to this method have changed.

CarolEidt replied:
Ah yes, though that wasn't me ;-) I'll update the header comment.

---
In reply to: [77410463](https://github.com/dotnet/coreclr/pull/7027#discussion_r77410463) [](ancestors = 77410463)

sivarv commented on src/jit/rationalize.h:
```diff
@@ -28,17 +28,16 @@ class Rationalizer : public Phase
-    // SIMD related transformations
-    void RewriteInitBlk(LIR::Use& use);
-    void RewriteCopyBlk(LIR::Use& use);
-    void RewriteObj(LIR::Use& use);
```
Glad to see these went away with 1st class structs.

CarolEidt commented:
> //    If we leave it as a GT_COPYBLK or GT_COPYOBJ we will call lvaSetVarDoNotEnregister() on both Source() and Dest()

Done. I also removed all the references to `GT_COPYBLK` and `GT_COPYOBJ` which no longer exist, and cleaned it up a bit more.

---
In reply to: [244479965](https://github.com/dotnet/coreclr/pull/7027#issuecomment-244479965) [](ancestors = 244479965)

---
Refers to: src/jit/morph.cpp:9198 in 86b72d1. [](commit_id = 86b72d15a213ddc964efc84043e50ec90339ba9b, deletion_comment = False)

CarolEidt commented:
>  *                       [dest]     [val/src]

Yes! will do.

---
In reply to: [244483265](https://github.com/dotnet/coreclr/pull/7027#issuecomment-244483265) [](ancestors = 244483265)

---
Refers to: src/jit/flowgraph.cpp:18337 in 86b72d1. [](commit_id = 86b72d15a213ddc964efc84043e50ec90339ba9b, deletion_comment = False)

sivarv commented:
> // we do some handling of contained immediates and widening operations of unsigneds

Since you are at it, please see if you can update function header.

---
Refers to: src/jit/lowerxarch.cpp:37 in 86b72d1. [](commit_id = 86b72d15a213ddc964efc84043e50ec90339ba9b, deletion_comment = False)

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -1442,24 +1459,56 @@ void Lowering::TreeNodeInfoInitCall(GenTreeCall* call)
+        }
+    }
 
+    if (isInitBlk)
```
This can simply be "else" block associated with if-stmnt on line 1481?

CarolEidt commented:
> // we do some handling of contained immediates and widening operations of unsigneds

I didn't make many changes here, but I added a simple function header.

---
In reply to: [244488941](https://github.com/dotnet/coreclr/pull/7027#issuecomment-244488941) [](ancestors = 244488941)

---
Refers to: src/jit/lowerxarch.cpp:37 in 86b72d1. [](commit_id = 86b72d15a213ddc964efc84043e50ec90339ba9b, deletion_comment = False)

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -5279,86 +5325,161 @@ void CodeGen::genConsumePutStructArgStk(
+//    sizeReg - The register into which the block's size should go
+//
+
+void CodeGen::genConsumeBlockSize(GenTreeBlk* blkNode, regNumber sizeReg)
```
Considering the magnitude of changes, i would suggest to subject these changes to the following additional test runs (both on windows and Linux). Note that these could be triggered by typing in PR comment window.

- corefx baseline
- Jitstress2
- gcstress

I am assuming that you will also be measuring throughput numbers to make sure that there is no regression.

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -5279,86 +5325,161 @@ void CodeGen::genConsumePutStructArgStk(
+        else
+        {
+            noway_assert(blkNode->gtOper == GT_STORE_DYN_BLK);
+            genConsumeRegAndCopy(blkNode->AsDynBlk()->gtDynamicSize, sizeReg);
```
Give the massive amount of changes, I cannot claim to have reviewed each line of code.  Based on what I have reviewed looks good to me.

sivarv commented:
:shipit:

briansull commented on src/jit/morph.cpp:
```diff
@@ -2045,12 +2045,9 @@ GenTreePtr Compiler::fgMakeTmpArgNode(
-                // Create an Obj of the temp to use it as a call argument.
-                arg = gtNewObjNode(lvaGetStruct(tmpVarNum), arg);
+                // Pass by value in two registers, using a regular GT_OBJ node, created below.
+                arg->gtFlags |= GTF_DONT_CSE;
```
I'm not sure why it would be necessary to set GTF_DONT_CSE on a GT_LCL_VAR node as such nodes are never considered CSE candidates.
If you understand why this method is setting this flag can you add a comment to the method header.

CarolEidt replied:
I will add a comment. This is to preserve existing behavior, but there should be a TODO-1stClassStructs to remove this.

---
In reply to: [77417012](https://github.com/dotnet/coreclr/pull/7027#discussion_r77417012) [](ancestors = 77417012)

briansull commented on src/jit/morph.cpp:
```diff
@@ -8499,44 +8501,88 @@ GenTreePtr Compiler::fgMorphOneAsgBlockOp(GenTreePtr tree)
+#if FEATURE_SIMD
+            if (varTypeIsSIMD(asgType))
+            {
+                assert(!isCopyBlock); // Else we would have returned the tree bove.
```
typo "bove" => above

briansull commented on src/jit/morph.cpp:
```diff
@@ -8804,64 +9249,90 @@ GenTreePtr Compiler::fgMorphCopyBlock(GenTreePtr tree)
+                if (destLclVar->lvType == TYP_STRUCT)
+                {
+                    // It would be nice if lvExactSize always corresponded to the size of the struct,
+                    // but it does not for the spill temps.
```
This could become a source of a lot of problems.
What does it mean to have a spill-temp for a struct?
we have lvExactSize which is ALWAUYS the exact size of the struct (like 6 bytes)
and lvSize() which is the rounded up size which would be 8.

CarolEidt replied:
These aren't register spill temps - they're the temps that the importer uses when it needs to spill the stack. I'm trying to reconstruct where that happens, but I'm having some difficulty now. This is not something new; it's something that I found was true for existing code as well. I'll add a "TODO" if that's OK, as I think it may take some time to track this down again.

---
In reply to: [77418893](https://github.com/dotnet/coreclr/pull/7027#discussion_r77418893) [](ancestors = 77418893)

briansull commented:
Based on what I have reviewed looks good to me.

briansull commented:
:shipit:

CarolEidt commented:
@dotnet-bot test Windows_NT jitstress2

CarolEidt commented:
@dotnet-bot test Windows_NT corefx_baseline

CarolEidt commented:
@dotnet-bot test Windows_NT gcstress0xc

benaadams commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build 
@dotnet-bot test Windows_NT x64 Checked Build and Test (Jit - CoreFx )

CarolEidt commented:
@dotnet-bot test Windows_NT corefx_baseline

CarolEidt commented:
@sivarv I did the additional tests you suggest. I did a perf run against ALL_FX, and the overall size was .01% larger (and should be actually smaller once I remove the requirement that all struct assignments have a block node). The throughput was about 1% faster, but the standard deviation was about twice that, so I would say it is a wash. The corefx build timed out. I'm starting it again.
cc @dotnet/jit-contrib

CarolEidt commented:
@sivarv the corefx test timed out twice. Given that the only asm diffs are SIMD tests, I feel pretty confident that this test won't expose new issues. So I'm going to go ahead and merge.
